cloudinit/config/cc_set_hostname.py

```diff
@@ -62,6 +65,11 @@ def handle(name, cfg, cloud, log, _args):
         log.debug(("Configuration option 'preserve_hostname' is set,"
                    " not setting the hostname in module %s"), name)
         return
+
+    # Set prefer_fqdn_over_hostname value in distro
+    if util.get_cfg_option_bool(cfg, "prefer_fqdn_over_hostname", False):
+        cloud.distro.set_option('prefer_fqdn_over_hostname', True)
```

Comment:
> There's also a `_select_hostname` in `distros/freebsd.py` that would need to be updated to account for the new config option.
> 
> Instead of having each distro override `_select_hostname`, I think it'd make more sense for the `Distro` base class in `distros/__init__.py` to have a `self.prefer_fqdn = False` member, and the two classes that default to true (rhel and freebsd) set it to true. Then the implementing classes don't need to override `_select_hostname` at all. Where we currently have code like `util.get_cfg_option_bool(self._cfg, "prefer_fqdn_over_hostname", False)`, it could instead be `util.get_cfg_option_bool(self._cfg, "prefer_fqdn_over_hostname", self.prefer_fqdn)`.
> 
> Does that make sense?

no not really sorry we are trying our best not to change default behaviour 
i did not override `_select_hostname` if u wanna do the above wee can merge this and me or u can do it in other PR

Reply:
> I'm honestly not sure how I feel about this. There's nothing particularly wrong with it, but it also feels like it's breaking an abstraction to one-off add this option to the distro configuration this way.
> 
> @blackboxsw or @OddBloke , do either of you have strong opinions here?

i think everything  that is shared between all distros should come parent distro ( like config here )

Reply:
```
diff --git a/cloudinit/distros/__init__.py b/cloudinit/distros/__init__.py
index 1b97342d..5d7e8535 100755
--- a/cloudinit/distros/__init__.py
+++ b/cloudinit/distros/__init__.py
@@ -77,8 +77,8 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
     # This is used by self.shutdown_command(), and can be overridden in
     # subclasses
     shutdown_options_map = {'halt': '-H', 'poweroff': '-P', 'reboot': '-r'}
-
     _ci_pkl_version = 1
+    prefer_fqdn = False
 
     def __init__(self, name, cfg, paths):
         self._paths = paths
@@ -257,12 +257,10 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
                         "hostname to %s", hostname)
 
     def _select_hostname(self, hostname, fqdn):
-        # Prefer the short hostname over the long
-        # fully qualified domain name
+        # Use the short hostname over the long fully qualified domain name
+        # unless manually overridden or a distro prefers the fqdn
         if util.get_cfg_option_bool(self._cfg, "prefer_fqdn_over_hostname",
-                                    False) and fqdn:
-            return fqdn
-        if not hostname:
+                                    self.prefer_fqdn) and fqdn:
             return fqdn
         return hostname
 
diff --git a/cloudinit/distros/freebsd.py b/cloudinit/distros/freebsd.py
index dde34d41..9659843f 100644
--- a/cloudinit/distros/freebsd.py
+++ b/cloudinit/distros/freebsd.py
@@ -27,12 +27,7 @@ class Distro(cloudinit.distros.bsd.BSD):
     pkg_cmd_remove_prefix = ["pkg", "remove"]
     pkg_cmd_update_prefix = ["pkg", "update"]
     pkg_cmd_upgrade_prefix = ["pkg", "upgrade"]
-
-    def _select_hostname(self, hostname, fqdn):
-        # Should be FQDN if available. See rc.conf(5) in FreeBSD
-        if fqdn:
-            return fqdn
-        return hostname
+    prefer_fqdn = True  # See rc.conf(5) in FreeBSD
 
     def _get_add_member_to_group_cmd(self, member_name, group_name):
         return ['pw', 'usermod', '-n', member_name, '-G', group_name]
diff --git a/cloudinit/distros/rhel.py b/cloudinit/distros/rhel.py
index 7d4d386e..6004eef3 100644
--- a/cloudinit/distros/rhel.py
+++ b/cloudinit/distros/rhel.py
@@ -49,6 +49,9 @@ class Distro(distros.Distro):
             }
         }
     }
+    # Should be fqdn if we can use it
+    # See: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/5/html/deployment_guide/ch-sysconfig  # noqa: E501
+    prefer_fqdn = True
 
     def __init__(self, name, cfg, paths):
         distros.Distro.__init__(self, name, cfg, paths)
@@ -91,17 +94,6 @@ class Distro(distros.Distro):
             }
             rhel_util.update_sysconfig_file(out_fn, host_cfg)
 
-    def _select_hostname(self, hostname, fqdn):
-        # Should be fqdn if we can use it
-        # See: https://www.centos.org/docs/5/html/Deployment_Guide-en-US/ch-sysconfig.html#s2-sysconfig-network # noqa
-        if not util.get_cfg_option_bool(self._cfg,
-                                        "prefer_fqdn_over_hostname",
-                                        True) and hostname:
-            return hostname
-        if fqdn:
-            return fqdn
-        return hostname
-
     def _read_system_hostname(self):
         if self.uses_systemd():
             host_fn = self.systemd_hostname_conf_fn

```
If you use `git apply` to apply ^, I think this won't change default behavior.

Reply:
i will apply the above